### PR TITLE
Feature/coroutines

### PR DESCRIPTION
**Pull request by n-balashov**
Homework #6 implementation

### CODE REVIEW

File: app/src/main/java/com/aacademy/homework/data/FakeDataRepository.kt [hunk 1]
```diff
@@ -3,27 +3,19 @@ package com.aacademy.homework.data
...
-        .flatMap {
-            if (it.isEmpty())
-                FakeApiRepository.getMovieDetail(id)
-            else
-                Single.just(it.first())
+    suspend fun getMovieDetail(id: Int): MovieDetailWithActors = FakeLocalRepository.getMovieDetail(id)
+        .let {
+            if (it.isEmpty()) FakeApiRepository.getMovieDetail(id) else it.first()
         }
 }
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/FakeDataRepository.kt]: Check formatting. You need have newline at end of every file.

File: app/src/main/java/com/aacademy/homework/data/api/FakeApiRepository.kt [hunk 2]
```diff
@@ -1,177 +1,27 @@
...
-            ).shuffled(Random(id * 100))
-        )
-    ).delay(2, SECONDS)
-        .map { MovieDetailWithActors(it, it.cast) }
-        .doOnSuccess {
-            FakeLocalRepository.cacheMovieDetailWithActors(it)
-        }
+    suspend fun getMovieDetail(id: Int): MovieDetailWithActors {
+        Thread.sleep(2000)
+        return loadMovieDetail(id, MyApp.INSTANCE)
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/api/FakeApiRepository.kt]: I'm not sure, but maybe better to pass the context as a parameter in the function.

File: app/src/main/java/com/aacademy/homework/data/local/FakeLocalRepository.kt [hunk 3]
```diff
@@ -16,23 +14,23 @@ object FakeLocalRepository {
             .build()
     }
 
-    fun getAllMoviePreviews(): Single<List<MoviePreviewWithTags>> {
+    suspend fun getAllMoviePreviews(): List<MoviePreviewWithGenres> {
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/local/FakeLocalRepository.kt]: You must write "withContext(Dispatchers.IO)" for the coroutine to be executed in IO

File: app/src/main/java/com/aacademy/homework/data/local/FakeLocalRepository.kt [hunk 4]
```diff
@@ -16,23 +14,23 @@ object FakeLocalRepository {
...
         return database.moviePreviewDao().getAllMovies()
     }
 
-    fun getMovieDetail(id: Int): Single<List<MovieDetailWithActors>> {
+    suspend fun getMovieDetail(id: Int): List<MovieDetailWithActors> {
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/local/FakeLocalRepository.kt]: withContext(Dispatchers.IO)

File: app/src/main/java/com/aacademy/homework/data/local/FakeLocalRepository.kt [hunk 5]
```diff
@@ -16,23 +14,23 @@ object FakeLocalRepository {
...
         return database.movieDetailDao().getMovieDetail(id)
     }
 
-    fun setMovieLiked(id: Int, isLiked: Boolean): Completable {
+    suspend fun setMovieLiked(id: Int, isLiked: Boolean) {
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/local/FakeLocalRepository.kt]: withContext(Dispatchers.IO)

File: app/src/main/java/com/aacademy/homework/data/local/FakeLocalRepository.kt [hunk 6]
```diff
@@ -16,23 +14,23 @@ object FakeLocalRepository {
...
         return database.moviePreviewDao().setMovieLiked(id, isLiked)
     }
 
-    fun cacheMoviePreviewsWithTags(moviePreviewsWithTags: List<MoviePreviewWithTags>) {
+    suspend fun cacheMoviePreviewsWithTags(moviePreviewsWithTags: List<MoviePreviewWithGenres>) {
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/local/FakeLocalRepository.kt]: withContext(Dispatchers.IO)

File: app/src/main/java/com/aacademy/homework/data/local/FakeLocalRepository.kt [hunk 7]
```diff
@@ -16,23 +14,23 @@ object FakeLocalRepository {
...
         database.moviePreviewDao().insert(moviePreviewsWithTags)
     }
 
-    fun cacheMovieDetailWithActors(movieDetailWithActors: MovieDetailWithActors) {
+    suspend fun cacheMovieDetailWithActors(movieDetailWithActors: MovieDetailWithActors) {
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/local/FakeLocalRepository.kt]: withContext(Dispatchers.IO)

File: app/src/main/java/com/aacademy/homework/data/local/dao/MovieDetailDao.kt [hunk 8]
```diff
@@ -9,31 +9,30 @@ import com.aacademy.homework.data.model.Actor
...
 
     @Transaction
     @Query("SELECT * FROM moviedetail WHERE id = :id")
-    fun getMovieDetail(id: Int): Single<List<MovieDetailWithActors>>
+    suspend fun getMovieDetail(id: Int): List<MovieDetailWithActors>
 
     @Transaction
     @Insert
-    fun insert(movieDetailWithActors: MovieDetailWithActors) {
+    suspend fun insert(movieDetailWithActors: MovieDetailWithActors) {
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/local/dao/MovieDetailDao.kt:22]: withContext(Dispatchers.IO)

File: app/src/main/java/com/aacademy/homework/data/local/dao/MoviePreviewDao.kt [hunk 9]
```diff
@@ -5,40 +5,38 @@ import androidx.room.Insert
...
-    fun getAllMovies(): Single<List<MoviePreviewWithTags>>
+    suspend fun getAllMovies(): List<MoviePreviewWithGenres>
 
     @Query("UPDATE moviepreview SET isLiked = :isLiked WHERE id = :id")
-    fun setMovieLiked(id: Int, isLiked: Boolean): Completable
+    suspend fun setMovieLiked(id: Int, isLiked: Boolean)
 
     @Transaction
-    fun insert(moviePreviewsWithTags: List<MoviePreviewWithTags>) {
+    suspend fun insert(moviePreviewsWithTags: List<MoviePreviewWithGenres>) {
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/local/dao/MoviePreviewDao.kt:24]: withContext(Dispatchers.IO)

File: app/src/main/java/com/aacademy/homework/data/model/MovieDetail.kt [hunk 10]
```diff
@@ -1,15 +1,10 @@
...
 data class MovieDetail(
     @PrimaryKey val id: Int,
-    val storyline: String,
-    @Ignore val cast: List<Actor>
-) {
-
-    constructor(id: Int, storyline: String) : this(id, storyline, emptyList())
-}
\ No newline at end of file
+    val overview: String,
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/model/MovieDetail.kt]: Why is there a comma?

File: app/src/main/java/com/aacademy/homework/data/model/MovieDetailWithActors.kt [hunk 11]
```diff
@@ -16,5 +16,5 @@ data class MovieDetailWithActors(
             entityColumn = "actorId"
         )
     )
-    val cast: List<Actor>,
+    val actors: List<Actor>,
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/model/MovieDetailWithActors.kt]: Why is there a comma?

File: app/src/main/java/com/aacademy/homework/data/model/MoviePreviewWithGenres.kt [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+        parentColumn = "id",
+        entity = Genre::class,
+        entityColumn = "id",
+        associateBy = Junction(
+            value = MovieTag::class,
+            parentColumn = "movieId",
+            entityColumn = "tagId"
+        )
+    )
+    val genres: List<Genre>,
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/model/MoviePreviewWithGenres.kt]: Comma

File: app/src/main/java/com/aacademy/homework/ui/activities/MainViewModel.kt [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+class MainViewModel : ViewModel() {
+
+    private val _moviesPreview = MutableLiveData<List<MoviePreviewWithGenres>>()
+    val moviesPreview: LiveData<List<MoviePreviewWithGenres>> = _moviesPreview
+
+    private val _movieDetail = MutableLiveData<MovieDetailWithActors>()
+    val movieDetail: LiveData<MovieDetailWithActors> = _movieDetail
+
+    init {
+        viewModelScope.launch {
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/ui/activities/MainViewModel.kt]: You should use viewModelScope.launch(Dispatchers.IO) else all coroutines would be run in Main thread especially if you don't add withContext in function

File: app/src/main/java/com/aacademy/homework/ui/activities/MainViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+    fun getMovieDetail(id: Int) {
+        if (movieDetail.value?.movieDetail?.id != id) {
+            viewModelScope.launch(Dispatchers.IO) {
+                _movieDetail.postValue(FakeDataRepository.getMovieDetail(id))
+            }
+        }
+    }
+
+    fun setMovieLiked(id: Int, isLiked: Boolean) {
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/ui/activities/MainViewModel.kt]: Dispatchers.IO

File: app/src/main/java/com/aacademy/homework/ui/moviedetail/FragmentMoviesDetails.kt [hunk 15]
```diff
@@ -39,73 +41,76 @@ class FragmentMoviesDetails : Fragment() {
...
         menu.clear()
         super.onCreateOptionsMenu(menu, inflater)
     }
 
-    override fun onDestroy() {
-        super.onDestroy()
-        compositeDisposable.dispose()
+    private fun initViews() {
+        binding.apply {
+            (activity as MainActivity?)?.let {
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/ui/moviedetail/FragmentMoviesDetails.kt]: https://kotlinlang.org/docs/reference/typecasts.html#unsafe-cast-operator

But you have one activity and (activity as MainActivity) would be safe

File: app/src/main/java/com/aacademy/homework/ui/movielist/FragmentMoviesList.kt [hunk 16]
```diff
@@ -6,26 +6,29 @@ import android.view.MenuInflater
...
 import com.aacademy.homework.utils.viewBinding
 import com.bumptech.glide.Glide
-import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
-import io.reactivex.rxjava3.disposables.CompositeDisposable
-import io.reactivex.rxjava3.schedulers.Schedulers
-import timber.log.Timber
 
 class FragmentMoviesList : Fragment(R.layout.fragment_movies_list) {
 
     private val binding by viewBinding(FragmentMoviesListBinding::bind)
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/ui/movielist/FragmentMoviesList.kt:21]: Use the same initialization everywhere

File: app/src/main/java/com/aacademy/homework/ui/movielist/FragmentMoviesList.kt [hunk 17]
```diff
@@ -6,26 +6,29 @@ import android.view.MenuInflater
...
-    private val compositeDisposable: CompositeDisposable by lazy { CompositeDisposable() }
 
-    private lateinit var movieAdapter: MovieAdapter
+    private val viewModel: MainViewModel by activityViewModels()
+
+    private val movieAdapter by lazy {
+        MovieAdapter(Glide.with(this), resources, {
+            (activity as MainActivity?)?.openMovieDetail(it)
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/ui/movielist/FragmentMoviesList.kt]: activity as MainActivity

File: app/src/main/java/com/aacademy/homework/ui/movielist/FragmentMoviesList.kt [hunk 18]
```diff
@@ -47,43 +50,28 @@ class FragmentMoviesList : Fragment(R.layout.fragment_movies_list) {
...
-            adapter = movieAdapter
-            itemAnimator = MovieItemAnimator()
-        }
-
-        ItemTouchHelper(DragManageAdapter(moveCallback = movieAdapter::swapItems)).attachToRecyclerView(
-            binding.rvMovies
-        )
+    private fun initViews() {
+        binding.apply {
+            (activity as MainActivity?)?.setSupportActionBar(toolbar)
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/ui/movielist/FragmentMoviesList.kt]: (activity as MainActivity)
or (activity as? MainActivity)

File: app/src/main/java/com/aacademy/homework/ui/movielist/MovieAdapter.kt [hunk 19]
```diff
@@ -83,18 +83,18 @@ class MovieAdapter(
...
+        fun bind(moviePreview: MoviePreviewWithGenres) {
             binding.tvName.text = moviePreview.moviePreview.title
-            glide.load(moviePreview.moviePreview.coverPath)
+            glide.load(moviePreview.moviePreview.poster)
                 .transition(DrawableTransitionOptions.withCrossFade())
                 .into(binding.ivCover)
             binding.tvAgeLimit.text =
                 resources.getString(string.ageLimitFormat).format(moviePreview.moviePreview.ageLimit)
-            binding.tvTags.text = moviePreview.tags.joinToString(", ") { it.name }
+            binding.tvTags.text = moviePreview.genres.joinToString(", ") { it.name }
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/ui/movielist/MovieAdapter.kt]: Maybe take only 3-4 tags?

image

File: app/src/main/res/layout/layout_cast_item.xml [hunk 20]
```diff
@@ -12,12 +12,12 @@
...
+        app:shapeAppearanceOverlay="@style/CastCard.ImageAppearance"
         tools:srcCompat="@tools:sample/avatars" />
 
     <TextView
         android:id="@+id/tvNameCast"
-        style="@style/SubText.LightGray.Bold"
+        style="@style/CastName"
         android:layout_width="@dimen/castPhotoSize"
         android:layout_height="wrap_content"
         android:layout_marginTop="6dp"
```
Comment (Pavel-Sidorovich) [inline app/src/main/res/layout/layout_cast_item.xml]: space1.5x

File: app/src/main/res/layout/layout_movie_item.xml [hunk 21]
```diff
@@ -119,11 +129,11 @@
...
             app:layout_constraintTop_toTopOf="@+id/tvReviews"
-            tools:rating="3" />
+            tools:rating="2" />
 
         <TextView
             android:id="@+id/tvReviews"
-            style="@style/CardText.DarkGray.Bold"
+            style="@style/CardMovieReview"
             android:layout_width="wrap_content"
             android:layout_height="12dp"
```
Comment (Pavel-Sidorovich) [inline app/src/main/res/layout/layout_movie_item.xml]: In dimens

File: app/src/main/res/values/styles.xml [hunk 22]
```diff
@@ -33,95 +24,108 @@
...
+        <item name="cornerFamily">rounded</item>
+        <item name="cornerSize">4dp</item>
     </style>
 
-    <style name="MainText.Pink.Normal" parent="MainText">
-        <item name="android:textColor">@color/pink</item>
+    <style name="MovieCard.ImageAppearance" parent="">
+        <item name="cornerFamily">rounded</item>
+        <item name="cornerSizeTopLeft">7dp</item>
+        <item name="cornerSizeTopRight">7dp</item>
```
Comment (Pavel-Sidorovich) [inline app/src/main/res/values/styles.xml]: Maybe also in dimens

File: app/src/main/res/values/styles.xml [hunk 23]
```diff
@@ -33,95 +24,108 @@
...
 
-    <style name="MainText.White.Normal" parent="MainText">
-        <item name="android:textColor">@color/white</item>
+    <style name="CardMovieMin" parent="CardText.Bold">
+        <item name="android:textColor">@color/darkDarkGray</item>
+    </style>
+
+    <style name="CardMovieName" parent="MainText.Bold">
+        <item name="android:textColor">@color/lightlightGray</item>
+        <item name="android:shadowColor">#80000000</item>
```
Comment (Pavel-Sidorovich) [inline app/src/main/res/values/styles.xml]: In colors? You often use

File: app/src/main/res/values/styles.xml [hunk 24]
```diff
@@ -33,95 +24,108 @@
...
     </style>
 
-    <style name="MainText.White.Bold" parent="MainText.Bold">
+    <style name="DetailMovieReviews" parent="MainText.Bold">
+        <item name="android:textColor">@color/darkGray</item>
+    </style>
+
+    <style name="DetailMovieStoryline" parent="MainText">
         <item name="android:textColor">@color/white</item>
+        <item name="android:maxLines">1000</item>
```
Comment (Pavel-Sidorovich) [inline app/src/main/res/values/styles.xml]: Why?

Comment (n-balashov) [reply app/src/main/res/values/styles.xml]: forgot to delete

File: app/src/main/java/com/aacademy/homework/data/api/JsonLoad.kt [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+@Serializable
+private class JsonGenre(val id: Int, val name: String)
+
+@Serializable
+private class JsonActor(
+    val id: Int,
+    val name: String,
+    @SerialName("profile_path")
+    val profilePicture: String
+)
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/aacademy/homework/data/api/JsonLoad.kt]: what the purpose of using 2 data classes for the same data and map them? ©nikslushkin

File: app/src/main/res/drawable/movie_card_bg.xml [hunk 26]
```diff
@@ -1,27 +1,8 @@
...
-            </gradient>
-        </aapt:attr>
-    </path>
-</vector>
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android">
+    <gradient
+        android:angle="90"
+        android:endColor="@color/colorMovieMaskEnd"
+        android:startColor="@color/colorMovieMaskStart"
```
Comment (Pavel-Sidorovich) [inline app/src/main/res/drawable/movie_card_bg.xml]: Maybe attribute?

File: app/src/main/res/values-night/themes.xml [hunk 27]
```diff
@@ -0,0 +1,29 @@
+<resources xmlns:tools="http://schemas.android.com/tools">
+    <!-- Base application theme. -->
+    <style name="Theme.AAHomeWork" parent="Theme.MaterialComponents.DayNight.NoActionBar">
+        <!-- Primary brand color. -->
+        <item name="colorPrimary">@color/colorBackground</item>
+        <item name="colorPrimaryDark">@color/colorBackground</item>
+
+        <!-- Status bar color. -->
+        <item name="android:statusBarColor" tools:targetApi="lollipop">@android:color/transparent</item>
```
Comment (Pavel-Sidorovich) [inline app/src/main/res/values-night/themes.xml]: https://developer.android.com/studio/write/tool-attributes#toolstargetapi

Your minimum API - 21

File: app/src/main/res/values-night/themes.xml [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+        <item name="colorAccentPink">@color/colorAccentPink</item>
+        <item name="colorText">@color/colorText</item>
+        <item name="colorCastNameText">@color/colorCastNameText</item>
+        <item name="colorSectionText">@color/colorSectionText</item>
+        <item name="colorTextShadow">@color/colorTextShadow</item>
+        <item name="colorReviewsText">@color/colorReviewsText</item>
+        <item name="colorMinText">@color/colorMinText</item>
+        <item name="colorMovieMaskStart">@color/colorMovieMaskStart</item>
+        <item name="colorMovieMaskEnd">@color/colorMovieMaskEnd</item>
+
```
Comment (Pavel-Sidorovich) [inline app/src/main/res/values-night/themes.xml]: Do you like this blank line?

File: app/src/main/java/com/aacademy/homework/data/local/dao/MovieDetailDao.kt [hunk 29]
```diff
@@ -1,13 +1,38 @@
...
+            insert(actor)
+            insert(MovieActor(movieDetailWithActors.movieDetail.id, actor.id))
+        }
+    }
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insert(movieDetail: MovieDetail): Long
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insert(actor: Actor): Long
```
Comment (nikslushkin) [inline app/src/main/java/com/aacademy/homework/data/local/dao/MovieDetailDao.kt]: let's remove return value

File: app/src/main/java/com/aacademy/homework/data/model/Actor.kt [hunk 30]
```diff
@@ -0,0 +1,11 @@
+package com.aacademy.homework.data.model
+
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+
+@Entity
+data class Actor(
+    @PrimaryKey val id: Int,
```
Comment (nikslushkin) [inline app/src/main/java/com/aacademy/homework/data/model/Actor.kt]: Long

File: app/src/main/java/com/aacademy/homework/data/model/Genre.kt [hunk 31]
```diff
@@ -0,0 +1,10 @@
+package com.aacademy.homework.data.model
+
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+
+@Entity
+data class Genre(
+    @PrimaryKey val id: Int,
```
Comment (nikslushkin) [inline app/src/main/java/com/aacademy/homework/data/model/Genre.kt]: Long

File: app/src/main/java/com/aacademy/homework/data/model/MovieDetail.kt [hunk 32]
```diff
@@ -1,10 +1,10 @@
-package com.aacademy.homework.data.local.model
+package com.aacademy.homework.data.model
 
 import androidx.room.Entity
 import androidx.room.PrimaryKey
 
 @Entity
 data class MovieDetail(
     @PrimaryKey val id: Int,
```
Comment (nikslushkin) [inline app/src/main/java/com/aacademy/homework/data/model/MovieDetail.kt]: Long

File: app/src/main/java/com/aacademy/homework/data/model/MoviePreview.kt [hunk 33]
```diff
@@ -0,0 +1,17 @@
+package com.aacademy.homework.data.model
+
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+
+@Entity
+data class MoviePreview(
+    @PrimaryKey val id: Int,
```
Comment (nikslushkin) [inline app/src/main/java/com/aacademy/homework/data/model/MoviePreview.kt]: Long

File: app/src/main/java/com/aacademy/homework/ui/activities/MainViewModel.kt [hunk 34]
```diff
@@ -0,0 +1,41 @@
...
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.aacademy.homework.data.FakeDataRepository
+import com.aacademy.homework.data.local.FakeLocalRepository
+import com.aacademy.homework.data.model.MovieDetailWithActors
+import com.aacademy.homework.data.model.MoviePreviewWithGenres
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+class MainViewModel : ViewModel() {
```
Comment (nikslushkin) [inline app/src/main/java/com/aacademy/homework/ui/activities/MainViewModel.kt]: `MoviesViewModel`

File: app/src/main/java/com/aacademy/homework/ui/movielist/MovieItemAnimator.kt [hunk 35]
```diff
@@ -86,32 +87,36 @@ class MovieItemAnimator : DefaultItemAnimator() {
...
         )
         fadeBackground.interpolator = DECELERATE_INTERPOLATOR
         fadeBackground.duration = 600
 
-        animatorSet.addListener(object : AnimatorListener {
-            override fun onAnimationStart(animator: Animator) {
-                dispatchRemoveStarting(holder)
-            }
+        animatorSet.addListener(
+            object : AnimatorListener {
```
Comment (nikslushkin) [inline app/src/main/java/com/aacademy/homework/ui/movielist/MovieItemAnimator.kt]: AnimatorListenerAdapter ?

File: app/src/main/res/drawable/movie_card_bg.xml (hunk 26 above)
Comment (n-balashov) [reply app/src/main/res/drawable/movie_card_bg.xml]: Forgot to remove this file

**Pull request merged**